src/lib/fragments/AgGrid.react.js

```diff
@@ -381,12 +442,74 @@ export default class DashAgGrid extends Component {
 
             propsToSet.virtualRowData = virtualRowData;
         }
-        propsToSet.columnState = this.state.gridColumnApi.getColumnState();
+        propsToSet.columnState = JSON.parse(
+            JSON.stringify(this.state.gridColumnApi.getColumnState())
+        );
         setProps(propsToSet);
     }
 
     componentDidMount() {
-        this.setState({mounted: true});
+        const {id} = this.props;
+        if (id) {
+            agGridRefs[id] = this.reference.current;
+            eventBus.dispatch(id);
+        }
+    }
+
+    componentWillUnmount() {
+        this.setState({mounted: false, gridApi: null, gridColumnApi: null});
+        if (this.props.id) {
+            delete agGridRefs[this.props.id];
+            eventBus.remove(this.props.id);
+        }
+    }
+
+    shouldComponentUpdate(nextProps, nextState) {
+        const {gridColumnApi, gridApi} = this.state;
+        const {columnState, filterModel, selectedRows} = nextProps;
+
+        if (
+            !equals(
+                {...omit(OMIT_PROP_RENDER, nextProps)},
+                {...omit(OMIT_PROP_RENDER, this.props)}
+            )
+        ) {
+            return true;
+        }
+        if (
+            !equals(
+                {...omit(OMIT_STATE_RENDER, nextState)},
+                {...omit(OMIT_STATE_RENDER, this.state)}
+            )
+        ) {
+            return true;
+        }
+        if (gridApi) {
+            if (columnState) {
+                if (
+                    !equals(
+                        columnState,
+                        JSON.parse(
+                            JSON.stringify(gridColumnApi.getColumnState())
+                        )
```

Comment:
I can't see the JSON cycle being necessary here, is it?
```suggestion
                        gridColumnApi.getColumnState()
```